halo2_proofs/src/poly/commitment.rs

```diff
@@ -81,21 +95,55 @@ impl<C: CurveAffine> Params<C> {
             g
         };
 
-        // Let's evaluate all of the Lagrange basis polynomials
-        // using an inverse FFT.
-        let mut alpha_inv = E::Scalar::ROOT_OF_UNITY_INV;
-        for _ in k..E::Scalar::S {
-            alpha_inv = alpha_inv.square();
-        }
+        println!("time1 : {:?}", Instant::now() - start_time);
+        let fast = true;
+
         let mut g_lagrange_projective = g_projective;
-        best_fft(&mut g_lagrange_projective, alpha_inv, k);
-        let minv = E::Scalar::TWO_INV.pow_vartime(&[k as u64, 0, 0, 0]);
-        parallelize(&mut g_lagrange_projective, |g, _| {
-            for g in g.iter_mut() {
-                *g *= minv;
+        if fast {
+            let mut root = E::Scalar::ROOT_OF_UNITY_INV;
+            root = root.invert().unwrap();
+            for _ in k..E::Scalar::S {
+                root = root.square();
             }
-        });
+            let mut o_power = vec![E::Scalar::zero(); n as usize];
+            parallelize(&mut o_power, |o_chunk, start| {
+                let mut current_o = root.pow_vartime(&[start as u64]);
+                for o_chunk_ptr in o_chunk.iter_mut() {
+                    *o_chunk_ptr = current_o;
+                    current_o *= root;
+                }
+            });
 
+            //let mut g_lagrange_projective = vec![E::G1::group_zero(); n as usize];
+            parallelize(&mut g_lagrange_projective, |g, start| {
+                //let mut current_g: E::G1 = g1.into();
+                //current_g *= s.pow_vartime(&[start as u64]);
+                for (idx, g) in g.iter_mut().enumerate() {
+                    let pos = idx + start;
+                    let xx = (o_power[pos] * E::Scalar::from(n).invert().unwrap())
+                        * (s.pow_vartime(&[n as u64]) - E::Scalar::one())
```

Comment:
How about computing and storing `E::Scalar::from(n).invert()` and `(s.pow_vartime(&[n as u64]) - E::Scalar::one())` outside the loop?